sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/CardinalityTest.java

```diff
@@ -221,7 +237,9 @@ void cardinalityLimits_asynchronousInstrument() {
     meter.counterBuilder("async-counter1").buildWithCallback(callback);
     meter.counterBuilder("async-counter2").buildWithCallback(callback);
 
-    assertThat(deltaReader.collectAllMetrics())
+    meterProvider.forceFlush().join(10, TimeUnit.SECONDS);
```

Comment:
I agree with the sentiment but find this to be a significantly worse UX when writing metrics tests. Can you confirm that users won't have to worry about flushing after we integrate metrics into our JUnit extension? If so then it seems fine to me

Reply:
Its going to depend on the type of test they're writing, but generally, yes integration of metrics into our JUnit extension will help. We could extend `OpenTelemetryExtension` with something like the snippet below, to flush metrics before returning them for callers to assert against. 
```
  public List<MetricData> getMetrics() {
    return getMetrics(true);
  }

  public List<MetricData> getMetrics(boolean forceFlush) {
    if (forceFlush) {
      openTelemetry.getSdkMeterProvider().forceFlush().join(10, TimeUnit.SECONDS);
    }
    return metricExporter.getFinishedMetricItems();
  }
```

If they're writing tests that depend on being able to configure non standard things like the clock, or multiple readers, then they won't be able to leverage `OpenTelemetryExtension` (as it stands today) since it doesn't provide any config options.

Question though: the sdk spec defines requirements for our sdk, but we include testing utilities like `InMemorySpanExproter` and `InMemoryMetricExporter` in the :sdk:testing` module, not the sdk. And we include a fair number of other convenience utilities which are not specified. By the same logic, I would think we're entitled to include `InMemoryMetricReader` as a testing utility in `:sdk:testing`. It certainly improve testing patterns so I would be in favor of this if allowed by the spec.

Reply:
Yeah thinking more, I don't think testing utils are an aspect of the spec. We need to provide testing utilities for helping with writing tests - including tests that don't have background threads polling for metrics :) So keeping InMemoryMetricReader in sdk-testing sounds better than removing it

Reply:
FYI - regarding #3943 if the way the spec is worded leads to us removing this class then we need to go reword the specification.

The Speicfication was written to allow a design decision of EITHER "Exporter" does everything *or* "Reader is Pull", "Exporter is Push".   

.NET AFAIK is the only impl of the first, most of the rest of us do the second.   I'm against this change because I think we should go clarify the specification, not remove the best utiltiy we have for testing the metrics SDK.

Reply:
How about just moving it to the test artifact, rather than deleting it?

Reply:
> How about just moving it to the test artifact, rather than deleting it?

Yeah sounds like we're all in favor of that.